File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
import sys
input = sys.stdin.readline

dx = [0, 0, -1, 1]
dy = [-1, 1, 0, 0]

def dfs(x, y):
    global count
    if x < 0 or x >= size or y < 0 or y >= size:
        return
    if Complex[x][y] == 1:
        count += 1
        Complex[x][y] = 0
        for i in range(4):
            xx = x + dx[i]
            yy = y + dy[i]
            dfs(xx, yy)

Complex = []
size = int(input())
for _ in range(size):
    Complex.append(list(map(int, input().strip())))
    
result = []
count = 0
for i in range(size):
    for j in range(size):
        if Complex[i][j] == 1:
            dfs(i, j)
            result.append(count)
            count = 0
            
result.sort()
print(len(result))
for k in result:
    print(k)

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.cc
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

int mapsize;
vector<vector<int>> complexMap;

int dx[] = {0, 0, -1, 1};
int dy[] = {-1, 1, 0, 0};

int count_per_complex;

void dfs(int x, int y) {
    if (x < 0 || x >= mapsize || y < 0 || y >= mapsize) {
        return;
    }
    
    if (complexMap[x][y] == 0) {
        return;
    }
    
    complexMap[x][y] = 0;
    count_per_complex++;
    
    for (int i = 0; i < 4; i++) {
        int next_x = x + dx[i];
        int next_y = y + dy[i];
        dfs(next_x, next_y);
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
    
    cin >> mapsize;
    
    complexMap.resize(mapsize, vector<int>(mapsize));
    for (int i = 0; i < mapsize; i++) {
        string row;
        cin >> row;
        for (int j = 0; j < mapsize; j++) {
            complexMap[i][j] = row[j] - '0';
        }
    }
    
    vector<int> result;
    
    for (int i = 0; i < mapsize; i++) {
        for (int j = 0; j < mapsize; j++) {
            if (complexMap[i][j] == 1) {
                count_per_complex = 0;
                dfs(i, j);
                result.push_back(count_per_complex);
            }
        }
    }
    
    sort(result.begin(), result.end());
    
    cout << result.size() << "\n";
    for (int count : result) {
        cout << count << "\n";
    }
    
    return 0;
}